python/ml4ir/applications/ranking/model/metrics/metrics_helper.py

```diff
@@ -2,143 +2,156 @@
 from typing import List, Union
 
+# Metrics where higher value is better/desirable
+POSITIVE_METRIC_SUFFIXES = [
+    "MRR",
+    "NDCG_mean"
+]
+# Metrics where lower value is better/desirable
+NEGATIVE_METRIC_SUFFIXES = [
+    "ACR",
+    "failure_all_mean",
+    "failure_any_mean",
+    "failure_all_rank_mean",
+    "failure_any_rank_mean",
+    "failure_any_count_mean",
+    "failure_any_fraction_mean"
+]
+
+
+def compute_secondary_label_metrics(
+    secondary_label_values: pd.Series,
+    ranks: pd.Series,
+    click_rank: int,
+    secondary_label: str,
+    prefix: str = "",
+):
+    """
+    Computes the secondary ranking metrics using a secondary label for a single query
 
-def compute_failure_stats(
-    df: pd.DataFrame,
-    query_key_col: str,
+    Parameters
+    ----------
+    secondary_label_values: pd.Series
+        Series object containing the secondary label values for a given query
+    ranks: pd.Series
+        Series object containing the ranks corresponding to the secondary label values
+    click_rank: int
+        Rank of the clicked record
+    secondary_label: str
+        Name of the secondary label used to define metric name
+    prefix: str
+        Prefix attached to the metric name
+
+    Returns
+    -------
+    dict
+        Key value pairs of the metric names and the associated computed values
+        using the secondary label
+    """
+    failure_all = 0
+    failure_any = 0
+    failure_count = 0
+    failure_fraction = 0.0
+
+    click_secondary_label_value = secondary_label_values[ranks == click_rank].values[0]
+    pre_click_secondary_label_values = secondary_label_values[ranks < click_rank]
+
+    if pre_click_secondary_label_values.size > 0:
+        # Query failure only if failure on all records
+        failure_all = (
+            1
+            if (pre_click_secondary_label_values < click_secondary_label_value).all()
+            else 0
+        )
+        # Query failure if failure on at least one record
+        failure_any = (
+            1
+            if (pre_click_secondary_label_values < click_secondary_label_value).any()
+            else 0
+        )
+        # Count of failure records
+        failure_count = (
+            pre_click_secondary_label_values < click_secondary_label_value
+        ).sum()
+        # Normalizing to fraction of potential records
+        failure_fraction = failure_count / (click_rank - 1)
+
+    return {
+        "{}{}_failure_all".format(prefix, secondary_label): failure_all,
+        "{}{}_failure_any".format(prefix, secondary_label): failure_any,
+        "{}{}_failure_all_rank".format(prefix, secondary_label): click_rank
+        if failure_all
+        else 0,
+        "{}{}_failure_any_rank".format(prefix, secondary_label): click_rank
+        if failure_any
+        else 0,
+        "{}{}_failure_any_count".format(prefix, secondary_label): failure_count,
+        "{}{}_failure_any_fraction".format(prefix, secondary_label): failure_fraction,
+    }
```

Comment:
This is not new code. I am simply removing it into a separate method, so that it is cleaner, unit testable and also easy to add new metrics like NDCG.